File: services/ai-gateway/app/prompts/grammar.py
"""语法学习 Prompt"""
from __future__ import annotations

GRAMMAR_SYSTEM = """你是一位资深的英语语法老师，擅长帮备考雅思/托福/GRE/SAT 的学员讲透语法。

任务：针对给定的语法知识点，写一份讲解 + 例句 + 练习题。

要求：
1. 讲解用中文，通俗易懂，重点讲「考点」和「易错点」
2. 给出 3-5 条核心规则，每条配一个英文例句和中文翻译
3. 出 {num} 道 4 选 1 选择题（1 正确 + 3 干扰），干扰项要合理
4. 每道题有中文解析，指出考点和为什么对/错

输出 JSON（仅 JSON）：
{{
  "explanation": "整体讲解（中文，100-200字）",
  "rules": [
    {{"rule": "规则说明（中文）", "example": "英文例句", "example_translation": "例句中文翻译"}}
  ],
  "examples": [
    {{"sentence": "英文例句", "translation": "中文翻译"}}
  ],
  "questions": [
    {{
      "question": "题目（英文）",
      "choices": [
        {{"label": "A", "text": "选项A"}},
        {{"label": "B", "text": "选项B"}},
        {{"label": "C", "text": "选项C"}},
        {{"label": "D", "text": "选项D"}}
      ],
      "correct_label": "A",
      "explanation": "中文解析（考点 + 对错原因）"
    }}
  ]
}}
"""


def build_grammar_messages(topic: str, title_en: str, num: int) -> list[dict]:
    system = GRAMMAR_SYSTEM.format(num=num)
    en = f"（英文：{title_en}）" if title_en else ""
    user = f"语法知识点：{topic}{en}\n\n请讲解并出题。"
    return [
        {"role": "system", "content": system},
        {"role": "user", "content": user},
    ]

File: apps/web/src/services/grammar.ts
/**
 * 语法学习 API
 */
import { config, fetchWithFallback } from '../config'

export interface GrammarTopic {
  id: string
  title: string
  titleEn: string
  description: string
  difficulty: string
  examTags: string[]
}

export interface GrammarChoice {
  label: string
  text: string
}

export interface GrammarRule {
  rule: string
  example: string
  example_translation: string
}

export interface GrammarExample {
  sentence: string
  translation: string
}

export interface GrammarQuestion {
  question: string
  choices: GrammarChoice[]
  correct_label: string
  explanation: string
}

export interface GrammarResult {
  topic: string
  title_en: string
  explanation: string
  rules: GrammarRule[]
  examples: GrammarExample[]
  questions: GrammarQuestion[]
  is_mock: boolean
}

export async function fetchGrammarTopics(): Promise<GrammarTopic[]> {
  const path = '/api/grammar/topics'
  const directUrl = config.aiGateway ? `${config.aiGateway}${path}` : path
  const resp = await fetchWithFallback(path, directUrl, { method: 'GET' })
  if (!resp.ok) throw new Error(`语法知识点加载失败 ${resp.status}`)
  const data = await resp.json()
  return data?.items ?? []
}

export async function generateGrammar(
  topic: string,
  titleEn: string,
  numQuestions: number = 4,
): Promise<GrammarResult> {
  const path = '/api/grammar/generate'
  const directUrl = config.aiGateway ? `${config.aiGateway}${path}` : path
  const resp = await fetchWithFallback(path, directUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ topic, title_en: titleEn, num_questions: numQuestions }),
  })
  if (!resp.ok) {
    const body = await resp.text().catch(() => '')
    throw new Error(`语法讲解生成失败 ${resp.status}: ${body.slice(0, 100)}`)
  }
  return resp.json()
}
